vncviewer/touch.cxx

```diff
@@ -0,0 +1,627 @@
+  int x_root, y_root;
+  XkbStateRec state;
+
+  // We don't have a real event to steal things from, so we'll have
+  // to fake these events based on the current state of things
+
+  root = XDefaultRootWindow(fl_display);
+  XTranslateCoordinates(fl_display, wnd, root,
+                        src.event_x,
+                        src.event_y,
+                        &x_root, &y_root, &child);
+  XkbGetState(fl_display, XkbUseCoreKbd, &state);
+
+  // XButtonEvent and XMotionEvent are almost identical, so we
+  // don't have to care which it is for these fields
+  dst->xbutton.serial = XLastKnownRequestProcessed(fl_display);
+  dst->xbutton.display = fl_display;
+  dst->xbutton.window = wnd;
+  dst->xbutton.root = root;
+  dst->xbutton.subwindow = None;
+  dst->xbutton.time = fl_event_time;
+  dst->xbutton.x = src.event_x;
+  dst->xbutton.y = src.event_y;
+  dst->xbutton.x_root = x_root;
+  dst->xbutton.y_root = y_root;
+  dst->xbutton.state = state.mods;
+  dst->xbutton.state |= ((state.ptr_buttons >> 1) & 0x1f) << 8; // FIXME: Check this!
+  dst->xbutton.same_screen = True; // FIXME
+}
+
+void FLTKGestureHandler::fakeMotionEvent(const XIDeviceEvent* origEvent)
+{
+  XEvent fakeEvent;
+
+  memset(&fakeEvent, 0, sizeof(XEvent));
+
+  fakeEvent.type = MotionNotify;
+  fakeEvent.xmotion.is_hint = False;
+  prepXEvent(&fakeEvent, origEvent);
+  //setTouchEventState(&fakeEvent);
+
+  fl_handle(fakeEvent);
+}
+
+void FLTKGestureHandler::fakeMotionEvent(const GHEvent origEvent)
+{
+  XEvent fakeEvent;
+  int mask;
+
+  memset(&fakeEvent, 0, sizeof(XEvent));
+
+  fakeEvent.type = MotionNotify;
+  fakeEvent.xmotion.is_hint = False;
+  prepXEvent(&fakeEvent, origEvent);
+  //setTouchEventState(&fakeEvent);
+  
+  switch (origEvent.gesture) {
+    case GH_LEFTBTN:
+      mask = Button1Mask;
+      break;
+    case GH_MIDDLEBTN:
+      mask = Button2Mask;
+      break;
+    case GH_RIGHTBTN:
+      mask = Button3Mask;
+      break;
+    //case GH_ZOOM:
+    //  mask = ControlMask;
+    //  break;
+    default:
+      mask = 0;
+  }
+
+  fakeEvent.xbutton.state |= mask;
+
+  fl_handle(fakeEvent);
+}
+
+void FLTKGestureHandler::fakeButtonEvent(bool press, int button,
+                                         const XIDeviceEvent* origEvent)
+{
+  XEvent fakeEvent;
+
+  memset(&fakeEvent, 0, sizeof(XEvent));
+
+  fakeEvent.type = press ? ButtonPress : ButtonRelease;
+  fakeEvent.xbutton.button = button;
+  prepXEvent(&fakeEvent, origEvent);
+  //setTouchEventState(&fakeEvent);
+
+  vlog.info("CORE xbutton.state = %i", fakeEvent.xbutton.state);
+
+  fl_handle(fakeEvent);
+}
+
+void FLTKGestureHandler::fakeButtonEvent(bool press, int button,
+                                         const GHEvent origEvent)
+{
+  XEvent fakeEvent;
+  int mask;
+
+  memset(&fakeEvent, 0, sizeof(XEvent));
+
+  fakeEvent.type = press ? ButtonPress : ButtonRelease;
+  fakeEvent.xbutton.button = button;
+  prepXEvent(&fakeEvent, origEvent);
+
+  switch (button) {
```

Comment:
Don't the mask and values have a strict relationship so you can just calculate the mask without a large switch statement?

Reply:
Yeah. I also did some testing with `xev` and it seems like the mask is only set for button release events, which I guess makes sense in a way. Anyway, see 461988b2